spec/jobs/application_job_spec.rb

```diff
@@ -2,34 +2,78 @@
-    described_class.queue_adapter = :resque
+    old_adapter = ActiveJob::Base.queue_adapter
+    ActiveJob::Base.queue_adapter = :resque
     example.run
-    described_class.queue_adapter = old_adapter
+    ActiveJob::Base.queue_adapter = old_adapter
   end
 
   it 'queues start empty' do
-    expect(Resque.info).to include(pending: 0)
+    expect(enqueued_jobs.size).to eq 0
   end
 
   context 'a subclass with message(s) queued' do
-    before do
-      allow(ZipmakerJob).to receive(:perform_later).and_call_original # undo rails_helper block
-      ZipmakerJob.perform_later('1234abc', 1)
-    end
-
     it 'does not add duplicate messages' do
-      expect { ZipmakerJob.perform_later('1234abc', 1) }
-        .not_to change { Resque.info[:pending] }.from(1)
+      RegularParameterJob.perform_later('1234abc', 1, false)
+      expect { RegularParameterJob.perform_later('1234abc', 1, false) }
+        .not_to change(enqueued_jobs, :size).from(1)
     end
 
     it 'but adds novel messages' do
-      expect { ZipmakerJob.perform_later('7890xyz', 1) } # different druid
-        .to change { Resque.info[:pending] }.from(1).to(2)
-      expect { ZipmakerJob.perform_later('1234abc', 2) } # same druid, different version
-        .to change { Resque.info[:pending] }.from(2).to(3)
+      RegularParameterJob.perform_later('1234abc', 1, false)
+      expect { RegularParameterJob.perform_later('7890xyz', 1, false) } # different druid
+        .to change(enqueued_jobs, :size).from(1).to(2)
+      expect { RegularParameterJob.perform_later('1234abc', 2, false) } # same druid, different version
+        .to change(enqueued_jobs, :size).from(2).to(3)
+    end
+
+    it 'cleans up its lock after succeeding so that the same job with the same params can be queued again immediately' do
+      RegularParameterJob.perform_later('4321cba', 1, false)
+
+      perform_enqueued_jobs
+
+      expect { RegularParameterJob.perform_later('4321cba', 1, false) }
+        .to change(enqueued_jobs, :size).from(0).to(1)
+    end
+
+    it 'cleans up its lock after failing so that the same job with the same params can be queued again immediately' do
+      RegularParameterJob.perform_later('0987zyx', 1, true)
+
+      begin
+        perform_enqueued_jobs
+      rescue # rubocop:disable Style/RescueStandardError terrible pattern in almost all cases, but acceptable here
+        # the error is bubbling up to here despite coming from an async job.  IRL, this would not happen.
+        # no idea why ActiveJob is behaving differently in the tests.
+      end
+
+      expect { RegularParameterJob.perform_later('0987zyx', 1, true) }
+        .to change(enqueued_jobs, :size).from(0).to(1)
+    end
+
+    it 'enqueues the job if it detects an existing lock that has expired' do
+      RegularParameterJob.perform_later('4321cba', 1, false)
+
+      sleep(RegularParameterJob.lock_timeout + 3) # lock timeout is very short, this should go well past it
```

Comment:
i always feel bad `sleep`ing in tests, but i looked and there's actually only one other `sleep` in the tests, for 0.1 seconds.  totally open to non-`sleep` suggestions for testing timeout.

Reply:
Seems the way around this would be to look for an existing lock, but if that code isn't already available to us (maybe it's private methods in resque or something), then meh.   I do hate sleeps, and I hate tests taking longer than they should ... but this can be a separate ticket.